src/util.cc

```diff
@@ -1,8 +1,196 @@
 #include "util.h"
 #include "string_bytes.h"
 
+#define UNI_SUR_HIGH_START    (uint32_t) 0xD800
+#define UNI_SUR_LOW_END       (uint32_t) 0xDFFF
+#define UNI_REPLACEMENT_CHAR  (uint32_t) 0x0000FFFD
+#define UNI_MAX_LEGAL_UTF32   (uint32_t) 0x0010FFFF
+
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4)
+#define HAS_GCC_BUILTIN_CLZ
+#endif
+
+#ifdef WIN32
+#include <intrin.h>
+static uint32_t __inline clz(uint32_t xs) {
+  unsigned long result = 0;
+  _BitScanReverse(&result, xs);
+  return (31 - result);
+}
+#elif defined(HAS_GCC_BUILTIN_CLZ)
+#define clz(xs) __builtin_clz(xs)
+#else
+static inline uint32_t clz(uint32_t xs) {
+  uint32_t out = 0;
+  while (xs >> (31 - out)) {
+    ++out;
+  }
+  return out;
+}
+#endif
```

Comment:
Compilers are pretty good in replacing idiomatic bit hacks with machine-specific instructions.  Have you checked whether it's really necessary to have these #ifdefs?